File: server/src/models/files.py
import os
import shutil
from urllib.parse import unquote

from models.tags import TagsDataModel


class FilesDataModel:
    def __init__(self, files_dir: list[str], tags_model: TagsDataModel):
        self.files_dir = files_dir
        self.tags_model = tags_model

    def get_files(self, sort=True):
        if os.path.exists(self.files_dir):
            # Get all files in files directory
            files = [
                f
                for f in os.listdir(self.files_dir)
                if os.path.isfile(os.path.join(self.files_dir, f))
            ]

            if sort:
                files.sort()

            return self.populate_files(files)

        return []

    def get_file_content(self, filename):
        file_path = os.path.join(self.files_dir, filename)

        if not os.path.exists(file_path):
            return None

        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()

        return {"name": filename, "content": content}

    def populate_files(self, files):
        tags = self.tags_model.get_tags()
        populated_files = []

        for file in files:
            file_info = self.get_file_content(filename=file)

            if not file_info:
                continue

            # Get tags for the file
            if file in tags:
                file_tags = tags[unquote(file)]
            else:
                file_tags = []

            populated_files.append(
                {
                    "name": file_info["name"],
                    "content": file_info["content"],
                    "tags": file_tags,
                }
            )

        return populated_files

    def add(self, filename, content):
        os.makedirs(self.files_dir, exist_ok=True)

        file_path = os.path.join(self.files_dir, filename)

        if os.path.exists(file_path):
            return None

        with open(file_path, "w") as f:
            f.write(content)

        return {"name": filename, "content": content}

    def add_buffer(self, filename, content_buffer):
        os.makedirs(self.files_dir, exist_ok=True)

        file_path = os.path.join(self.files_dir, filename)

        if os.path.exists(file_path):
            return None

        with open(file_path, "wb") as buffer:
            shutil.copyfileobj(content_buffer, buffer)

        return {
            "name": filename,
        }

    def update(self, filename, content):
        os.makedirs(self.files_dir, exist_ok=True)

        file_path = os.path.join(self.files_dir, filename)

        if not os.path.exists(file_path):
            return None

        with open(file_path, "w") as f:
            f.write(content)

        return {"name": filename}

    def delete(self, filename):
        os.makedirs(self.files_dir, exist_ok=True)

        file_path = os.path.join(self.files_dir, filename)

        if not os.path.exists(file_path):
            return None

        os.remove(file_path)

File: server/src/models/files.js
import sqlite3 from 'sqlite3';
import { open } from 'sqlite';
import { v4 as uuidv4 } from 'uuid';

class FilesDataModel {
    constructor(dbPath, tagsModel) {
        this.tagsModel = tagsModel;
        this.dbPromise = open({
            filename: dbPath,
            driver: sqlite3.Database
        }).then(async (db) => {
            await db.exec(`
                CREATE TABLE IF NOT EXISTS files (
                    id TEXT PRIMARY KEY,
                    name TEXT UNIQUE,
                    content TEXT,
                    created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
                )
            `);
            return db;
        }).catch((error) => {
            console.error('[files] Error opening files database:', error);
        });
    }

    async getFiles(sort = false) {
        const db = await this.dbPromise;

        try {
            // Join files and tags, aggregate tags as array
            const rows = await db.all(`
                SELECT 
                    f.id, 
                    f.name, 
                    f.content, 
                    f.created_at,
                    GROUP_CONCAT(t.tag, '|||') as tags
                FROM files f
                LEFT JOIN tags t ON t.file_id = f.id
                GROUP BY f.id
            `);

            let files = rows.map(row => ({
                id: row.id,
                name: row.name,
                content: row.content,
                tags: row.tags ? row.tags.split('|||') : []
            }));

            if (sort) {
                files = files.sort((a, b) => new Date(b.created_at) - new Date(a.created_at));
            }

            return files;
        } catch (e) {
            console.error('[files] Error fetching files:', e);
            return [];
        }
    }

    async getFileContent(identifier) {
        const db = await this.dbPromise;

        try {
            const row = await db.get('SELECT content FROM files WHERE id = ?', identifier);

            return row ? row.content : null;
        } catch (e) {
            console.error('[files] Error fetching file content:', e);
            return null;
        }
    }

    async add(filename, content) {
        const db = await this.dbPromise;
        const id = uuidv4();

        try {
            await db.run('INSERT INTO files (id, name, content) VALUES (?, ?, ?)', id, filename, content);
            return { id, name: filename, content };
        } catch (e) {
            console.error('[files] Error adding file:', e);
            return null;
        }
    }

    async update(identifier, filename, content) {
        const db = await this.dbPromise;

        try {
            const row = await db.get('SELECT id FROM files WHERE id = ?', identifier);

            if (!row) {
                return null;
            }

            await db.run('UPDATE files SET name = ?, content = ? WHERE id = ?', filename, content, row.id);
            return { id: row.id, name: filename, content };
        } catch (e) {
            console.error('[files] Error updating file:', e);
            return null;
        }
    }

    async delete(identifier) {
        const db = await this.dbPromise;

        try {
            const row = await db.get('SELECT id, name FROM files WHERE id = ?', identifier);

            if (!row) {
                return null;
            }

            await db.run('DELETE FROM files WHERE id = ?', row.id);
            return row.name;
        } catch (e) {
            console.error('[files] Error deleting file:', e);
            return null;
        }
    }
}

export default FilesDataModel;